### PR TITLE
feat(rome_js_parser): instantiation expressions #3147

### PR DESCRIPTION
**Pull request by denbezrukov**
## Summary
This PR fixes: 
https://github.com/rome/tools/issues/3793 
https://github.com/rome/tools/issues/3147

When a potential type argument list is followed by

- a line break,
- an ( token,
- a template literal string, or
- any token except < or > that isn't the start of an expression,

Typescript reference: 

- https://github.com/microsoft/TypeScript/blob/212d3864ef08a78454b2a12f7e1f11789917440e/src/compiler/parser.ts#L6394-L6413

Typescript design notes:

- https://github.com/microsoft/TypeScript/issues/49384
- https://github.com/microsoft/TypeScript/issues/49444

Relative issues:

- https://github.com/microsoft/TypeScript/pull/49353
- https://github.com/microsoft/TypeScript/issues/49362
- https://github.com/babel/babel/pull/14457
- https://github.com/evanw/esbuild/pull/2188

Open questions:

- Typescript has an extra condition in `isBinaryOperator` function: https://github.com/microsoft/TypeScript/blob/main/src/compiler/parser.ts#L5528-L5534
- Typescript and babel dissalow expression with type parameters as left side of property access [Prettier playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEqAPACoB8AFAJT7AC+AOlEwPQuGlkB0ARgNxNi5ANoByHqIC6fEABoQEAA65oAZ2SgAhgCdtEAO4AFHQnUpNAG32aAnuvk9tmsAGs4MAMqaAtnAAyOFBwyKiWqnAOTq7uHorOgQDmyDDa6BEgcN48cAAmObl+mlAJ6JoJcABiENremjC4xcggmugwEHIgABYw3hYA6p048KpxYHAepkM4AG5DNk1gqvYggeHaMIZOCbUhYekAVqoAHh6JFnAAiugQ8LsW4fJx2mtNMDaKcKpg2jjKHYo-WB9HA5GCdZAADgADI89OE+k5FE0AZ84NppsF5ABHa7wTZKMzNVQAWiCuVyHW0cBxOCpmzKOyQoXu6XC3hwyVSrLOcAAgvUfjxWnBDGiAkE7g8QKoeVcbsEmXt5DBNDxgaDwUgAEzKpw4CyJADCEG8jIyqgArB10OESKqzMypdM0gBJKD5WAeb6-GC890eN7nSVwOh0IA):
https://github.com/microsoft/TypeScript/pull/49464
https://github.com/babel/babel/pull/14650
- `f<T> << f<T>;` is valid code. But this PR doesn't support it. May be we can do it in another PR. [Prettier playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzAPAFQHwAJ3q4Y4DcIANCBAA4wCW0AzsqAIYBO7EA7gAocJmKVgBturAJ7NKAI3aswAazgwAyqwC2cADJ0ocZKlGM4s+UpWrqCvQHNkMdgFdTIOBplwAJl+-bWULZOrLZwAGIQ7BqsMPSByCCsTjAQFCAAFjAaIgDq6XTwjNZgcKqCBXQAbgUSCWCM0iB6JuwwvPK20YbGrgBWjAAeqnYicACKThDw3SImlNbsLQkwEtRwjGDsdLRp1FuwOXReMOnIABwADPNcJjny1Al763DslQaUAI6T8O00QomMAC0+m83jS7DgXzoEPaIS6SCMs1cJg0dAczmRIzgAEFYlsZMk4LwXrp9DM5iBGFiJlMDAiepQYKwZIdjqckAAmRnyOgiOwAYQgGnhbkYAFY0k4TJhmUJERTKi4AJJQXywVSbbYwbGq1QrUbkuAAXyNQA)

## Test Plan

`cargo test -p rome_js_parser`

test cases from babel:

- https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/typescript/type-arguments/instantiation-expression-asi
- https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/typescript/type-arguments/instantiation-expression-binary-operator
- https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/typescript/type-arguments/instantiation-expression-false-positive-2
- https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/typescript/type-arguments/instantiation-expression-false-positive-3
- https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/typescript/type-arguments/instantiation-expression-false-positive-4
- https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/typescript/type-arguments/instantiation-expression-false-positive
- https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/typescript/type-arguments/instantiation-expression-optional-chain
- https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/typescript/type-arguments/instantiation-expression-property-access
- https://github.com/babel/babel/tree/main/packages/babel-parser/test/fixtures/typescript/type-arguments/instantiation-expression

test cases from esbuild:

- https://github.com/evanw/esbuild/blob/895f50c84b47f2506c02b74972da9640db800c54/internal/js_parser/ts_parser_test.go#L1916

### CODE REVIEW

File: crates/rome_js_parser/src/syntax/typescript/types.rs [hunk 1]
```diff
@@ -1294,35 +1516,25 @@ pub(crate) fn parse_ts_type_arguments_in_expression(p: &mut JsParser) -> ParsedS
...
+        // by `>` is ambiguous with a (re-scanned) `>>` operator, so we disqualify both. Also, in
+        // this context, `+` and `-` are unary operators, not binary operators.
+        T![<] | T![>] | T![+] | T![-] => false,
+        // We favor the type argument list interpretation when it is immediately followed by
+        // a line break, a binary operator, or something that can't start an expression.
+        _ => p.has_preceding_line_break() || is_binary_operator(p) || !is_at_expression(p),
+    }
+}
+
+fn is_binary_operator(p: &mut JsParser) -> bool {
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/syntax/typescript/types.rs:1534]: Nit

```suggestion
fn is_at_binary_operator(p: &JsParser) -> bool {
```

I recommend moving this to `expr.rs` next to the binary expression parsing

File: crates/rome_js_parser/src/syntax/typescript/types.rs [hunk 2]
```diff
@@ -1294,35 +1516,25 @@ pub(crate) fn parse_ts_type_arguments_in_expression(p: &mut JsParser) -> ParsedS
...
-        | T![!==]
-        | T![&&]
-        | T![||]
-        | T![??]
-        | T![^]
-        | T![&]
-        | T![|]
-        | T!['}']
-        | EOF
+        OperatorPrecedence::try_from_binary_operator(p.cur()),
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/syntax/typescript/types.rs:1536]: You can check if `in` is a valid operator by accepting the `ExpressionContext` and then use `context.is_in_included()` (`true` -> `in` is a binary expression, `false` -> `in` is not a binary expression)


I guess a test case would be

```
for (const a = b.test<string> in []) {
 
}
```

This code is invalid but it's important that `in` isn't handled as binary expression to correctly parse this as a `for ... in` loop

File: crates/rome_js_parser/src/syntax/typescript/types.rs [hunk 3]
```diff
@@ -1256,19 +1256,241 @@ fn parse_ts_type_predicate(p: &mut JsParser) -> ParsedSyntax {
...
 // type StringBox = Box<string>;
 
-// test_err ts ts_instantiation_expressions1
+// test ts ts_instantiation_expressions_1
+// class A {
+//  constructor() {
+//    f<T> super();
+//  }
+// }
+// f<TemplateStringsArray>``;
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/syntax/typescript/types.rs:1285]: TheUniverseTimAndEricMindBlownGIF (2)

File: crates/rome_js_parser/src/syntax/typescript/types.rs [hunk 4]
```diff
@@ -1256,19 +1256,241 @@ fn parse_ts_type_predicate(p: &mut JsParser) -> ParsedSyntax {
 // const b4 = f<number>?.<number>();  // Type Error, expected no type arguments
 // const x1 = f<true>
 // (true);
-// // Parsed as relational expression
+// // Parsed as instantiation expression
 // const x2 = f<true>
 // true;
 // // Parsed as instantiation expression
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/syntax/typescript/types.rs:1262]:
```suggestion
```

File: crates/rome_js_parser/test_data/inline/ok/ts_instantiation_expressions.ts [hunk 5]
```diff
@@ -12,9 +12,19 @@ const b3 = f<number>?.();
 const b4 = f<number>?.<number>();  // Type Error, expected no type arguments
 const x1 = f<true>
 (true);
-// Parsed as relational expression
+// Parsed as instantiation expression
 const x2 = f<true>
 true;
 // Parsed as instantiation expression
```
Comment (MichaReiser) [inline crates/rome_js_parser/test_data/inline/ok/ts_instantiation_expressions.ts:18]:
```suggestion
```

**Comment by MichaReiser** [comment]
Let's rebase this PR to get the latest parser conformance tests. This PR should fix some of them.

File: crates/rome_js_parser/test_data/inline/ok/ts_instantiation_expressions.ts (hunk 5 above)
Comment (denbezrukov) [reply crates/rome_js_parser/test_data/inline/ok/ts_instantiation_expressions.ts]: removed

File: crates/rome_js_parser/src/syntax/typescript/types.rs (hunk 4 above)
Comment (denbezrukov) [reply crates/rome_js_parser/src/syntax/typescript/types.rs]: removed

File: crates/rome_js_parser/src/syntax/typescript/types.rs (hunk 2 above)
Comment (denbezrukov) [reply crates/rome_js_parser/src/syntax/typescript/types.rs]: added new test case and logic with context.

File: crates/rome_js_parser/src/syntax/typescript/types.rs (hunk 1 above)
Comment (denbezrukov) [reply crates/rome_js_parser/src/syntax/typescript/types.rs]: moved

**Review by MichaReiser — CHANGES_REQUESTED** [review]
I ran the typescript instantiation expression tests and there's one test that is passing when it should not:

* `[FAIL] type-arguments\instantiation-expression-property-access: Incorrectly passed`

The test should fail because it's invalid to access a property after an instantiation expression.

```javascript
// invalid
a<b>.c;
a<b>?.c;
a<b>?.[c];
```

I assume that TypeScript implements this as part of the type checker and not as part of the parser itself (but it would be worth double-checking). 

We have two options if that's the case:

* Try to implement it as part of the parser
* Implement a syntax rule that reports property accessors after instantiation expressions

And we can implement this either as part of this or a separate PR. Let me know what you prefer.

**Comment by denbezrukov** [comment]
Babel ts parser and typescript parser have a different implementation about access a property after an instantiation expression. 

- https://github.com/microsoft/TypeScript/pull/49464
- https://github.com/babel/babel/pull/14650

We want to have the same behaviour as typescript parser. [Typescript playground](https://www.typescriptlang.org/play?ssl=14&ssc=11&pln=13&pc=11#code/PTAEDcEMBsEsBMBQiBmAeARgPgPwDoAKAYwEoBuVTXPAbSIF0KD1sS8imWsT8PFmqPWgwrIQoWADsocJF3YUuvCpHzoiWPPEVpImgOZUyQA)

Added new test cases and an error.

**Review by MichaReiser — APPROVED** [review]

**Comment by MichaReiser** [comment]
Thank you so much for working on this. Well done!

**Pull request merged**